File: egg-niuke-training/04-intersection-of-two-linked-lists.cpp
/**
 * NC66 两个链表的第一个公共节点
 * https://www.nowcoder.com/practice/6ab1d9a29e88450685099d45c9e31e46
 * 
 * 输入两个无环的单向链表，找出它们的第一个公共节点。
 * 
 * 时间复杂度：O(m+n)
 * 空间复杂度：O(1)
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    /**
     * 找到第一个公共节点 - 双指针法
     */
    ListNode* FindFirstCommonNode(ListNode* head1, ListNode* head2) {
        if (!head1 || !head2) {
            return nullptr;
        }
        
        ListNode* p1 = head1;
        ListNode* p2 = head2;
        
        // 两个指针分别走完两个链表
        while (p1 != p2) {
            p1 = p1 ? p1->next : head2;
            p2 = p2 ? p2->next : head1;
        }
        
        return p1;
    }
};

int main() {
    // 创建两个相交的链表
    ListNode* common = new ListNode(6);
    common->next = new ListNode(7);
    
    ListNode* head1 = new ListNode(1);
    head1->next = new ListNode(2);
    head1->next->next = new ListNode(3);
    head1->next->next->next = common;
    
    ListNode* head2 = new ListNode(4);
    head2->next = new ListNode(5);
    head2->next->next = common;
    
    Solution solution;
    ListNode* result = solution.FindFirstCommonNode(head1, head2);
    
    if (result) {
        cout << "第一个公共节点值: " << result->val << endl;
    } else {
        cout << "无公共节点" << endl;
    }
    
    return 0;
}

File: egg-niuke-training/04-intersection-of-two-linked-lists.py
"""
NC66 两个链表的第一个公共节点
https://www.nowcoder.com/practice/6ab1d9a29e88450685099d45c9e31e46

输入两个无环的单向链表，找出它们的第一个公共节点，如果没有公共节点则返回空。

解法：双指针
让两个指针分别走完两个链表，这样就能在第二轮相遇在公共节点。

时间复杂度：O(m+n)
空间复杂度：O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def find_first_common_node(head1, head2):
    """
    找到第一个公共节点 - 双指针法
    """
    if not head1 or not head2:
        return None
    
    p1, p2 = head1, head2
    
    # 两个指针分别走完两个链表
    # p1走完链表1再走链表2，p2走完链表2再走链表1
    # 这样走的总长度相同，会在公共节点相遇
    while p1 != p2:
        p1 = p1.next if p1 else head2
        p2 = p2.next if p2 else head1
    
    return p1

# 测试
if __name__ == "__main__":
    # 创建两个相交的链表
    # 链表1: 1 -> 2 -> 3
    #                     \
    #                      6 -> 7
    #                     /
    # 链表2:      4 -> 5
    
    common = ListNode(6, ListNode(7))
    
    head1 = ListNode(1, ListNode(2, ListNode(3, common)))
    head2 = ListNode(4, ListNode(5, common))
    
    result = find_first_common_node(head1, head2)
    if result:
        print(f"第一个公共节点值: {result.val}")
    else:
        print("无公共节点")

